registry/quilt_server/auth.py

```diff
@@ -0,0 +1,419 @@
+    uuid = generate_uuid()
+    token = Token(user_id=user_id, token=uuid)
+    db.session.add(token)
+
+    exp = exp or get_exp()
+    payload = {'id': user_id, 'uuid': uuid, 'exp': exp}
+    token = jwt.encode(payload, app.secret_key, algorithm='HS256')
+    return token.decode('utf-8')
+
+def consume_code_string(code_str):
+    code = decode_code(code_str)
+    return consume_code(code['id'], code['code'])
+
+def consume_code(user_id, code):
+    found = Code.query.filter_by(user_id=user_id).with_for_update().one_or_none()
+    if found is None:
+        return None
+    if found.token != code:
+        return None
+    db.session.delete(code)
+    return user_id
+
+def verify_hash(password, pw_hash):
+    try:
+        if not pwd_context.verify(password, pw_hash):
+            raise ApiException(401, 'Password verification failed')
+    except ValueError:
+        raise ApiException(401, 'Password verification failed')
+
+def try_login(username, password):
+    user = User.get_by_name(username)
+    if not user:
+        return False
+
+    if not user.is_active:
+        return False
+
+    try:
+        verify_hash(password, user.password)
+    except ApiException:
+        return False
+    update_last_login(user)
+    return True
+
+def create_admin():
+    # Only runs in dev -- only dev_config reads in the environment variables
+    try:
+        admin_username = app.config['DEV_USERNAME']
+        admin_password = app.config['DEV_PASSWORD']
+        admin_email = app.config['DEV_EMAIL']
+    except KeyError:
+        return
+    if not admin_username or not admin_password or not admin_email:
+        return
+    _create_or_update_user(admin_username, password=admin_password, email=admin_email,
+                 is_admin=True, is_active=True)
+    db.session.commit()
+
+app.before_first_request(create_admin)
+
+
+linkgenerator = itsdangerous.URLSafeTimedSerializer(
+    app.secret_key,
+    salt='quilt'
+    )
+
+def dump_link(payload, salt=None):
+    link = linkgenerator.dumps(payload, salt=salt)
+    return link.replace('.', '~')
+
+def load_link(link, max_age, salt=None):
+    payload = link.replace('~', '.')
+    return linkgenerator.loads(payload, max_age=max_age, salt=salt)
+
+ACTIVATE_SALT = 'activate'
+PASSWORD_RESET_SALT = 'reset'
+MAX_LINK_AGE = 60 * 60 * 24 # 24 hours
+
+def generate_activation_token(user_id):
+    existing_token = ActivationToken.get(user_id)
+    new_token = existing_token or ActivationToken(user_id=user_id, token=generate_uuid())
+    db.session.add(new_token)
+    return new_token.token
+
+def consume_activation_token(user_id, token):
+    found = ActivationToken.get(user_id)
+    if not found:
+        return False
+    if found.token != token:
+        return False
+    db.session.delete(found)
+    return True
+
+def generate_reset_token(user_id):
+    existing_token = PasswordResetToken.get(user_id)
+    reset_token = existing_token or PasswordResetToken(user_id=user_id, token=generate_uuid())
```

Comment:
You can force a race condition here:
- Add `import time; time.sleep(1)` here
- Run `flask run --with-threads` (ok, in prod, we're using processes, not threads, but doesn't matter)
- Make sure you don't already have a password reset token
- Run `for i in `seq 10`; do curl 'http://localhost:5000/reset_password' -H 'content-type: application/json' --data '{"email":"dima@quiltdata.io"}' & done` (with your username)

You now have ten or so reset tokens. Try to actually reset the password using one of the links. You'll get:
```
  File "/home/dima/src/quilt/registry/quilt_server/views.py", line 387, in reset_password_endpoint
    return reset_password_response()
  File "/home/dima/src/quilt/registry/quilt_server/auth.py", line 64, in reset_password_response
    payload = verify_reset_link(link)
  File "/home/dima/src/quilt/registry/quilt_server/auth.py", line 408, in verify_reset_link
    if not consume_reset_token(payload['id'], payload['token']):
  File "/home/dima/src/quilt/registry/quilt_server/auth.py", line 376, in consume_reset_token
    found = PasswordResetToken.get(user_id)
  File "/home/dima/src/quilt/registry/quilt_server/models.py", line 261, in get
    return cls.query.filter(cls.user_id == user_id).one_or_none()
  File "/home/dima/envs/quilt-server3/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2827, in one_or_none
    "Multiple rows were found for one_or_none()")
sqlalchemy.orm.exc.MultipleResultsFound: Multiple rows were found for one_or_none()
```

Reply:
(In fact, once this happens, you can't recover: old reset password links don't work, and you can't create a new one, either.)